doc/content/xcp-networkd/host-network-device-ordering.md

```diff
@@ -0,0 +1,181 @@
+network device (with the same MAC address) has changed since the last ordering.
+The current unordered network device is considered newly added.
+3. Found with a MAC address, and no current devices have this MAC address: This
+indicates that a new network device has replaced the old one in the same PCI
+slot.
+
+The replacing network device should be assigned the same position as the
+replaced one.
+
+Removed devices
+---------------
+
+A network device can be removed or unplugged since the last ordering. Its
+position, MAC address, and PCI address are saved for future reference, and its
+position will be reserved. This means there may be a gap in the order: a
+position that was previously assigned to a network device is now vacant because
+the device has been removed.
+
+Newly added devices
+-------------------
+
+As long as `the assumption` holds, newly added devices since the last ordering
+can be assigned positions consistently across multiple hosts. Newly added
+devices will not be assigned the positions reserved for removed devices.
+
+Removed and then added back
+---------------------------
+
+It is a supported scenario for a removed device to be plugged back in,
+regardless of whether it is in the same PCI slot or not. This can be recognized
+by searching for the device in the saved removed devices using its MAC address.
+The reserved position will be reassigned to the device when it is added back.
+
+Multinic functions
+------------------
+
+The multinic function is a special kind of network device. When this type of
+physical device is plugged into a PCI slot, multiple network devices are
+reported at a single PCI address. Additionally, the number of reported network
+devices may change due to driver updates.
+
+As long as `the assumption` holds, the initial order of these devices can be
+generated automatically and kept stable by using MAC addresses to identify
+individual devices. However, `biosdevname` cannot reliably generate an order for
+all devices reported at one PCI address. For devices located at the same PCI
+address, their MAC addresses are used to generate the initial order.
+
+For example, suppose `biosdevname` generates an order for a multinic function
+and other non-multinic devices. Within this order, the N devices of the
+multinic function with MAC addresses mac[1], ..., mac[N] are assigned positions
+pos[1], ..., pos[N] correspondingly. `biosdevname` cannot ensure that the device
+with mac[1] is always assigned position pos[1]. Instead, it ensures that the
+entire set of positions pos[1], ..., pos[N] remains stable for the devices of
+the multinic function. Therefore, to ensure the order follows the MAC address
+order, the devices of the multinic function need to be sorted by their MAC
+addresses within the set of positions.
+
+Rare cases that can not be handled automatically
+------------------------------------------------
+
+In summary, to keep the order stable, the auto-generated order needs to be saved
```

Comment:
How much of a problem are two NICs with identical PCI slot? How does this arise? Is this a card with multiple ethernet ports? Or can you really have two physical cards with the same PCI slot (I would be surprised).

Reply:
The `two NICs with identical PCI slot` is not rare case. The rare cases are the ones that the sorting logic's output doesn't match with the user's expectation.
The sorting logic uses MAC or PCI slot to track the changes. But it may fail. E.g. a user may unplug a device from a PCI slot and plug it to another slot. The current logic thinks the network of the NIC doesn't change. But what the user want is remove the NIC and just use the NIC for a new network. And after a few host boot up, a new NIC is plugged as replacement for the original network.
The sorting logic can only handle limited cases.